### PR TITLE
Strongly typed handler option and standardize mappers a bit more

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

- add targeting option for directories named with platforms
- add a PropertyMapper version that allows for strongly typed handlers to reduce casting inside mappers
- create some variations of options for how to create mappers/extensions methods to see what makes people happy

### Testing Procedure ###
- be happy with what gets merged

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: .nuspec/Xamarin.Forms.MultiTargeting.targets [hunk 1]
```diff
@@ -2,22 +2,32 @@
 	<ItemGroup Condition="$(TargetFramework.StartsWith('Xamarin.iOS')) != true ">
 		<Compile Remove="**\*.iOS.cs" />
 		<None Include="**\*.iOS.cs" />
+    <Compile Remove="**\iOS\*.cs" />
```
Comment (PureWeen) [inline .nuspec/Xamarin.Forms.MultiTargeting.targets:5]: Right now for our platform folders we have directories named after the platforms and then files in those folders with "extensions" 

These allow for a project to separate by folder name if they want to

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Button/ButtonHandler.Android.cs [hunk 2]
```diff
@@ -9,9 +9,9 @@ public partial class ButtonHandler : AbstractViewHandler<IButton, AppCompatButto
 	{
 		protected override AppCompatButton CreateView() => new AppCompatButton(Context);
 
-		public static void MapText(IViewHandler handler, IButton view) 
+		public static void MapText(ButtonHandler handler, IButton view) 
 		{
-			((AppCompatButton)handler.NativeView).Text = view.Text;
+			handler.TypedNativeView.Text = view.Text;
```
Comment (PureWeen) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Button/ButtonHandler.Android.cs:14]: Yay less casting!!!

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Label/LabelHandler.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+#if __IOS__
+using NativeView = UIKit.UILabel;
```
Comment (PureWeen) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Label/LabelHandler.cs:3]: Is this cool? Annoying? 

This way we basically have on handler that just creates the default functions and then the extension methods are where the behavior splits

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Mac/ButtonExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using AndroidX.AppCompat.Widget;
+
+namespace Xamarin.Platform
+{
+	public static class ButtonExtensions
+	{
+		public static void UpdateColor(this IButton button, NSButton button)
```
Comment (PureWeen) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Mac/ButtonExtensions.cs:10]: Notice ButtonExtensions.cs doesn't have ".mac."  

It's sort of nice having ".mac." for navigation purposes but it seems a bit excessive to have /mac/file.mac.cs

File: src/Platform.Handlers/samples/Sample.Droid/MainActivity.cs [hunk 5]
```diff
@@ -35,7 +35,7 @@ void Add(params IView[] views)
 		{
 			foreach (var view in views)
 			{
-				_page.AddView(view.ToNative(this), new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WrapContent, ViewGroup.LayoutParams.WrapContent));
+				_page.AddView(view.ToNative(this));
```
Comment (PureWeen) [inline src/Platform.Handlers/samples/Sample.Droid/MainActivity.cs:38]: These wrap content calls mess up the Arrange calls

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/ActionMapper.cs [hunk 6]
```diff
@@ -1,22 +1,24 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Text;
+using Xamarin.Platform.Handlers;
 
 namespace Xamarin.Platform
 {
-	public class ActionMapper<TVirtualView>
+	public class ActionMapper<TVirtualView, TViewHandler>
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/ActionMapper.cs:8]: don't you want a constraint on TViewHandler being IViewHandler ?

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Label/LabelHandler.cs (hunk 3 above)
Comment (StephaneDelcroix) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Label/LabelHandler.cs:3]: annoying, but a necessary evil.

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs:1]: with the new MultiTarget.targets, do you really need the `Android` folder AND the `.Android.cs` suffix ?

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+using System.Collections.Generic;
+using System.Text;
+using AndroidX.AppCompat.Widget;
+
+namespace Xamarin.Platform
+{
+	public static class ButtonExtensions
+	{
+		public static void UpdateColor(this AppCompatButton appCompatButton, IButton button)
+		{
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs:11]: are we ready to discuss style ? if so, we're moving to c#8, I propose that we move to expression-body methods for those extensions. It really sends the message "don't bother, it's just a one-liner boilerplate helper"

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 9]
```diff
@@ -64,8 +64,9 @@ public virtual (Action<IViewHandler, IFrameworkElement> Action, bool RunOnUpdate
 		public virtual IReadOnlyList<string> UpdateKeys => updateKeys ?? PopulateKeys(ref updateKeys);
 	}
 
-	public class PropertyMapper<TVirtualView> : PropertyMapper, IEnumerable
+	public class PropertyMapper<TVirtualView, TViewHandler> : PropertyMapper, IEnumerable
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:67]: type constraint ?

**Review by StephaneDelcroix — APPROVED** [review]

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Label/LabelHandler.cs (hunk 3 above)
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Label/LabelHandler.cs:3]: Yea that was the general consensus among the team :-)

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/ActionMapper.cs (hunk 6 above)
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/ActionMapper.cs:8]: https://github.com/xamarin/Xamarin.Forms/pull/12198/files/e2b03332fad9781efafe5b5b853d275374a262f7#diff-d7ba16578d6228dae25138978651ce28R10

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs (hunk 9 above)
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:67]: https://github.com/xamarin/Xamarin.Forms/pull/12198/files/e2b03332fad9781efafe5b5b853d275374a262f7#diff-5df2195417becc796212aec09601b894R69

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs (hunk 7 above)
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs]: No, I removed the suffix on these files

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs (hunk 8 above)
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/ButtonExtensions.Android.cs]: Converted these to expression-body

**Pull request merged**